### PR TITLE
Fix #383 Pull replication not going back into IDLE state

### LINKED ISSUE #383: Pull replication not going back into IDLE state
To reproduce:

* Run TodoLite-Android
* Do a pull sync
* Update a document locally

**Actual**: spinny will never stop spinning
**Expected**: it should stop spinning after a short duration

I suspect it may have been caused by https://github.com/couchbase/couchbase-lite-java-core/issues/299

### PR DESCRIPTION
**Pull request by hideki**

### CODE REVIEW

**Comment by tleyden** [comment]
@pasin can you take a look and pull me in if there's anything I can help with?

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java [hunk 1]
```diff
@@ -845,63 +845,45 @@ public void run() {
...
-
                     try {
-
                         if (batcher != null) {
-                            Log.d(Log.TAG_SYNC, "batcher.waitForPendingFutures()");
+                            // if batcher delays task execution, need to wait same amount of time. (0.5 sec or 0 sec)
+                            try { Thread.sleep(batcher.delayToUse()); }catch(Exception e){}
+                            Log.e(Log.TAG_SYNC, "batcher.waitForPendingFutures()");
                             batcher.waitForPendingFutures();
                         }
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Can we refactor waitForPendingFutures to take that delay into account as the batcher knows that it has something pending? I understand that ScheduledFuture.get() method should already take the delay into account or I misunderstand that.

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java [hunk 2]
```diff
@@ -845,63 +845,45 @@ public void run() {
...
-                        Log.d(Log.TAG_SYNC, "waitForPendingFutures()");
+                        Log.e(Log.TAG_SYNC, "waitForPendingFutures()");
                         waitForPendingFutures();
 
                         if (downloadsToInsert != null) {
-                            Log.d(Log.TAG_SYNC, "downloadsToInsert.waitForPendingFutures()");
+                            // if batcher delays task execution, need to wait same amount of time. (1.0 sec or 0 sec)
+                            try { Thread.sleep(downloadsToInsert.delayToUse()); }catch(Exception e){}
+                            Log.e(Log.TAG_SYNC, "downloadsToInsert.waitForPendingFutures()");
                             downloadsToInsert.waitForPendingFutures();
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Same as above, can we make waitForPendingFutures to take delay into account?

File: src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java [hunk 3]
```diff
@@ -1102,21 +1102,21 @@ protected void initializeStateMachine() {
...
+                Log.v(Log.TAG_SYNC, "called notifyChangeListenersStateTransition");
             }
         });
 
         stateMachine.configure(ReplicationState.RUNNING).onExit(new Action1<Transition<ReplicationState, ReplicationTrigger>>() {
             @Override
             public void doIt(Transition<ReplicationState, ReplicationTrigger> transition) {
-                Log.v(Log.TAG_SYNC, "[onExit()] " + transition.getSource() + " => " + transition.getDestination());
-                Log.d(Log.TAG_SYNC, "replicator exiting the RUNNING method");
+                Log.v(Log.TAG_SYNC, "[onEntry()] " + transition.getSource() + " => " + transition.getDestination());
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java]: Should the log message tagged with 'onExit()'?

File: src/main/java/com/couchbase/lite/support/Batcher.java [hunk 4]
```diff
@@ -167,6 +166,17 @@ public int count() {
...
+     * this is mainly for debugging
+     */
+    public int sizeOfPendingFutures() {
+        synchronized(this) {
+            if(pendingFutures == null) {
+                return 0;
+            }
+            return pendingFutures.size();
+        }
+    }
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/support/Batcher.java]: Are we using this in uni tests or an hoc debugging? I think we just clean it out if it's just for one-time debugging.

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java [hunk 5]
```diff
@@ -845,63 +845,45 @@ public void run() {
...
-                try {
-                    Log.d(Log.TAG_SYNC, "changeTrackerCaughtUp");
-                } catch (Exception e) {
-                    e.printStackTrace();
-                    throw new RuntimeException(e);
-                }
-            }
-        });
-
+        Log.e(Log.TAG_SYNC, "changeTrackerCaughtUp");
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Should this be something else other than Log.e? I understand that Log.e is for error.

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java [hunk 6]
```diff
@@ -845,63 +845,45 @@ public void run() {
...
                         if (batcher != null) {
-                            Log.d(Log.TAG_SYNC, "batcher.waitForPendingFutures()");
+                            // if batcher delays task execution, need to wait same amount of time. (0.5 sec or 0 sec)
+                            try { Thread.sleep(batcher.delayToUse()); }catch(Exception e){}
+                            Log.e(Log.TAG_SYNC, "batcher.waitForPendingFutures()");
                             batcher.waitForPendingFutures();
                         }
 
-                        Log.d(Log.TAG_SYNC, "waitForPendingFutures()");
+                        Log.e(Log.TAG_SYNC, "waitForPendingFutures()");
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Same as above, should this be Log.e?

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java [hunk 7]
```diff
@@ -845,63 +845,45 @@ public void run() {
...
                         waitForPendingFutures();
 
                         if (downloadsToInsert != null) {
-                            Log.d(Log.TAG_SYNC, "downloadsToInsert.waitForPendingFutures()");
+                            // if batcher delays task execution, need to wait same amount of time. (1.0 sec or 0 sec)
+                            try { Thread.sleep(downloadsToInsert.delayToUse()); }catch(Exception e){}
+                            Log.e(Log.TAG_SYNC, "downloadsToInsert.waitForPendingFutures()");
```
Comment (pasin) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Same as above, should this be Log.e?

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java (hunk 1 above)
Comment (hideki) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Problem is that inbox in Batcher is not empty, but pendingFeatures in Batcher is empty. Calling `batcher.waitForPendingFutures();` before `Feature` is put in pendingFeatures does not make sense. 

We might be able to change `Batcher.waitForPendingFutures()` method to wait `inbox.size()` becomes 0 before calling `ScheduledFuture.get() `.

What do you think?

File: src/main/java/com/couchbase/lite/replicator/PullerInternal.java (hunk 2 above)
Comment (hideki) [inline src/main/java/com/couchbase/lite/replicator/PullerInternal.java]: Same as above

**Pull request merged**